internal/contour/annotations.go

```diff
@@ -76,6 +78,33 @@ func parseAnnotationUInt32(annotations map[string]string, annotation string) *ty
 	return &types.UInt32Value{Value: uint32(v)}
 }
 
+// parseUpstreamProtocols parses the annotations map for a contour.heptio.com/upstream-protocol.{protocol}
+// where 'protocol' identifies which protocol must be used in the upstream.
+// If the value is not present, or malformed, then an empty map is returned.
+func parseUpstreamProtocols(annotations map[string]string, annotation string, protocols []string) map[string]struct{} {
+	up := make(map[string]struct{})
+	for _, protocol := range protocols {
+		ports, ok := annotations[fmt.Sprintf("%s.%s", annotation, protocol)]
+		if !ok {
+			continue
+		}
+
+		for _, port := range strings.Split(ports, ",") {
+			if port != "" {
```

Comment:
you'll have to use a strings.Trim here because someone might write `"80, 443"`

Reply:
OMG, beginner failure. Will do.